gc.man
.TH BDWGC 3 "23 Aug 2023"
.SH NAME
MANAGED_STACK_ADDRESS_BOEHM_GC_malloc, MANAGED_STACK_ADDRESS_BOEHM_GC_malloc_atomic, MANAGED_STACK_ADDRESS_BOEHM_GC_free, MANAGED_STACK_ADDRESS_BOEHM_GC_realloc, MANAGED_STACK_ADDRESS_BOEHM_GC_enable_incremental,
MANAGED_STACK_ADDRESS_BOEHM_GC_register_finalizer, MANAGED_STACK_ADDRESS_BOEHM_GC_malloc_ignore_off_page,
MANAGED_STACK_ADDRESS_BOEHM_GC_malloc_atomic_ignore_off_page, MANAGED_STACK_ADDRESS_BOEHM_GC_set_warn_proc \- Garbage collecting
malloc replacement
.SH SYNOPSIS
#include <gc/gc.h>
.br
void * MANAGED_STACK_ADDRESS_BOEHM_GC_malloc(size_t size);
.br
void * MANAGED_STACK_ADDRESS_BOEHM_GC_malloc_atomic(size_t size);
.br
void MANAGED_STACK_ADDRESS_BOEHM_GC_free(void *ptr);
.br
void * MANAGED_STACK_ADDRESS_BOEHM_GC_realloc(void *ptr, size_t size);
.br
void MANAGED_STACK_ADDRESS_BOEHM_GC_enable_incremental(void);
.br
void * MANAGED_STACK_ADDRESS_BOEHM_GC_malloc_ignore_off_page(size_t size);
.br
void * MANAGED_STACK_ADDRESS_BOEHM_GC_malloc_atomic_ignore_off_page(size_t size);
.br
void MANAGED_STACK_ADDRESS_BOEHM_GC_set_warn_proc(void (*proc)(char *, MANAGED_STACK_ADDRESS_BOEHM_GC_word));
.br
.sp
cc ... -lgc
.LP
.SH DESCRIPTION
.I MANAGED_STACK_ADDRESS_BOEHM_GC_malloc
and
.I MANAGED_STACK_ADDRESS_BOEHM_GC_free
are plug-in replacements for standard malloc and free.  However,
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_malloc
will attempt to reclaim inaccessible space automatically by invoking
a conservative garbage collector at appropriate points.  The collector
traverses all data structures accessible by following pointers from the
machines registers, stack(s), data, and bss segments.  Inaccessible structures
will be reclaimed.  A machine word is considered to be a valid pointer if
it is an address inside an object allocated by
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_malloc
or friends.
.LP
In most cases it is preferable to call the macros MANAGED_STACK_ADDRESS_BOEHM_GC_MALLOC, MANAGED_STACK_ADDRESS_BOEHM_GC_FREE, etc.
instead of calling MANAGED_STACK_ADDRESS_BOEHM_GC_malloc and friends directly.  This allows debugging
versions of the routines to be substituted by defining MANAGED_STACK_ADDRESS_BOEHM_GC_DEBUG before
including gc.h.
.LP
See the documentation in the include files gc_cpp.h and gc_allocator.h,
as well as the gcinterface.md file in the distribution,
for an alternate, C++ specific interface to the garbage collector.
Note that C++ programs generally
need to be careful to ensure that all allocated memory (whether via new,
malloc, or STL allocators) that may point to garbage collected memory
is either itself garbage collected, or at least traced by the collector.
.LP
Unlike the standard implementations of malloc,
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_malloc
clears the newly allocated storage.
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_malloc_atomic
does not.  Furthermore, it informs the collector that the resulting object
will never contain any pointers, and should therefore not be scanned by the
collector.
.LP
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_free
can be used to deallocate objects, but its use is optional, and generally
discouraged.
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_realloc
has the standard realloc semantics.  It preserves pointer-free-ness.
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_register_finalizer
allows for registration of functions that are invoked when an object becomes
inaccessible.
.LP
The garbage collector tries to avoid allocating memory at locations that
already appear to be referenced before allocation.  (Such apparent
``pointers'' are usually large integers and the like that just happen to look
like an address.)  This may make it hard to allocate very large objects.
An attempt to do so may generate a warning.
.LP
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_malloc_ignore_off_page
and
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_malloc_atomic_ignore_off_page
inform the collector that the client code will always maintain a pointer to
near the beginning (i.e. within the first heap block) of the object, and that
pointers beyond that can be ignored by the collector.  This makes it much
easier for the collector to place large objects.  These are recommended for
large object allocation.  (Objects expected to be > ~100 KB should be
allocated this way.)
.LP
It is also possible to use the collector to find storage leaks in programs
destined to be run with standard malloc/free.  The collector can be compiled
for thread-safe operation.  Unlike standard malloc, it is safe to call malloc
after a previous malloc call was interrupted by a signal, provided the
original malloc call is not resumed.
.LP
The collector may, on rare occasion, produce warning messages.  On UNIX
machines these appear on stderr.  Warning messages can be filtered,
redirected, or ignored with
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_set_warn_proc
This is recommended for production code.  See gc.h for details.
.LP
Fully portable code should call
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_INIT
from the primordial thread of the main program before making any other
GC calls.  On most platforms this does nothing and the collector is
initialized on first use.  On a few platforms explicit initialization is
necessary.  And it can never hurt.
.LP
Debugging versions of many of the above routines are provided as macros.
Their names are identical to the above, but consist of all capital letters.
If MANAGED_STACK_ADDRESS_BOEHM_GC_DEBUG is defined before gc.h is included, these routines do additional
checking, and allow the leak detecting version of the collector to produce
slightly more useful output.  Without MANAGED_STACK_ADDRESS_BOEHM_GC_DEBUG defined, they behave exactly
like the lower-case versions.
.LP
On some machines, collection will be performed incrementally after a call to
.I
MANAGED_STACK_ADDRESS_BOEHM_GC_enable_incremental.
This may temporarily write protect pages in the heap.  See the README file for
more information on how this interacts with system calls that write to the
heap.
.LP
Other facilities not discussed here include limited facilities to support
incremental collection on machines without appropriate VM support, provisions
for providing more explicit object layout information to the garbage
collector, more direct support for ``weak'' pointers, support for
``abortable'' garbage collections during idle time, etc.
.LP
.SH "SEE ALSO"
The README and gc.h files in the distribution.  More detailed definitions of
the functions exported by the collector are given there.  (The above list is
not complete.)
.LP
The web site at http://www.hboehm.info/gc/ (or https://github.com/ivmai/bdwgc/).
.LP
Boehm, H., and M. Weiser, "Garbage Collection in an Uncooperative Environment",
"Software Practice & Experience", September 1988, pp. 807-820.
.LP
The malloc(3) man page.
.LP
.SH AUTHOR
Hans-J. Boehm (boehm@acm.org).
Some of the code was written by others (see the AUTHORS file for the details),
most notably by Alan Demers, and, recently, Ivan Maidanski.


ChangeLog

== [8.3.0] (development) ==

* Acquire GC lock in MANAGED_STACK_ADDRESS_BOEHM_GC_is_tmp_root and MANAGED_STACK_ADDRESS_BOEHM_GC_register_disclaim_proc
* Add 'lib' prefix to build artifact names in Makefile.direct
* Add API to get total stopped-world marking time
* Add I_HOLD_LOCK assertion to scratch_alloc/recycle and their callers
* Add I_HOLD_LOCK assertions to ensure no race on reveal hidden pointers
* Add Linux example to README.cmake
* Add assertion about GC lock to MANAGED_STACK_ADDRESS_BOEHM_GC_add_roots_inner and its callers
* Add assertion about kind argument range in MANAGED_STACK_ADDRESS_BOEHM_GC_register_describe_type_fn
* Add assertion that GC lock is held in start/continue_reclaim, reclaim_all
* Add assertion that custom setspecific is not called to overwrite value
* Add assertion to MANAGED_STACK_ADDRESS_BOEHM_GC_maybe_gc that GC is not in progress
* Add assertion to ensure MANAGED_STACK_ADDRESS_BOEHM_GC_unmapped_bytes cannot underflow
* Add assertion to verify MANAGED_STACK_ADDRESS_BOEHM_GC_threads hash table entries are chained properly
* Add assertions about MANAGED_STACK_ADDRESS_BOEHM_GC_mark_state to alloc.c and mark.c
* Add assertions about hidden ptr lowest bit in toggle-refs implementation
* Add callback API to support ephemeron marking
* Add check that gc_hdrs.h and gc_locks.h are included only from gc_priv.h
* Add check-deps goal to Makefiles to build the tests but not execute them
* Add debug logging to MANAGED_STACK_ADDRESS_BOEHM_GC_clear_exclusion_table and MANAGED_STACK_ADDRESS_BOEHM_GC_exclude_static_roots
* Add minimal testing of MANAGED_STACK_ADDRESS_BOEHM_GC_allow_register_threads
* Add missing cast in MANAGED_STACK_ADDRESS_BOEHM_GC_is_init_called (refactoring)
* Add missing undef MANAGED_STACK_ADDRESS_BOEHM_GC_NO_THREAD_REDIRECTS in threaded tests
* Add support of wasm32-wasi target
* Add testing of remaining API functions (gctest and leaktest)
* Add verbose logging of events of getting memory from OS
* Adjust comments in block_unmap_inner
* Adjust comments referring glibc version (documentation)
* Adjust header includes in cord .c files (refactoring)
* Adjust indent inside case statements in configure.ac (reformatting)
* Adjust indentation in C preprocessor directives (reformatting)
* Adjust naming of Win32/64 and x86/64 words in comments and documentation
* Adjust pointer-to-int cast in MANAGED_STACK_ADDRESS_BOEHM_GC_register_dynamic_libraries for WinCE
* Adjust printing on amount of unmapped memory after GC
* Allow MANAGED_STACK_ADDRESS_BOEHM_GC_push_roots to work even when no exclusions registered
* Allow custom AO_SRC_DIR, CURSES, EXEEXT, HOSTCC/CFLAGS for Makefile.direct
* Allow disable USE_PROC_FOR_LIBRARIES even if malloc is redirected on Linux
* Allow gc_allocator<T> and friends to be defined in namespace boehmgc
* Allow incremental GC on OpenBSD
* Allow mprotect-based incremental GC on AIX and for all Linux targets
* Allow passing custom CC and CXX to Makefile.direct
* Allow passing custom CFLAGS[_EXTRA] to Makefile.direct
* Allow single signal usage for both world suspend and resume in gctest
* Allow testing of retry-signals functionality
* Allow to build with forced WRAP_MARK_SOME
* Allow to compile tests with custom NTHREADS set to zero
* Allow to randomly choose a CPU core if AO ops are emulated with locks
* Allow to skip heap expansion in MANAGED_STACK_ADDRESS_BOEHM_GC_init
* Allow to unmap memory block right in GC cycle where block is freed
* Allow to use print/check_marks_int_list() in gctest
* Allow to use same signal for thread suspend and restart (default on E2K)
* Always invoke pending finalizers in MANAGED_STACK_ADDRESS_BOEHM_GC_generic_malloc_uncollectable
* Always print error details message when gctest fails
* Always set only one bit past end in MANAGED_STACK_ADDRESS_BOEHM_GC_set_hdr_marks
* Assert SET_HDR does not write to MANAGED_STACK_ADDRESS_BOEHM_GC_all_nils
* Assert in MANAGED_STACK_ADDRESS_BOEHM_GC_typed_mark_proc that GC lock is held by some collecting thread
* Assert that bm_table is already initialized when MANAGED_STACK_ADDRESS_BOEHM_GC_double_descr used
* Assert that disappearing links are accessible during registration and GC
* Assert that size of element of a typed object is non-zero (refactoring)
* Avoid code duplication between pthread_support.c and win32_threads.c
* Avoid code duplication in MANAGED_STACK_ADDRESS_BOEHM_GC_core_gcj_malloc (refactoring)
* Avoid code duplication in MANAGED_STACK_ADDRESS_BOEHM_GC_make_array_descriptor (refactoring)
* Avoid code duplication in IGNORE_OFF_PAGE-specific malloc functions
* Avoid code duplication regarding MANAGED_STACK_ADDRESS_BOEHM_GC_push_roots in MANAGED_STACK_ADDRESS_BOEHM_GC_mark_some_inner
* Avoid duplication of magic numbers in disclaim.c, weakmap.c (refactoring)
* Avoid unmap of memory allocated during GC init regardless of threshold
* Better document MANAGED_STACK_ADDRESS_BOEHM_GC_generate_random_heap/valid_address API functions
* Better document MANAGED_STACK_ADDRESS_BOEHM_GC_no_dls and NO_DEBUGGING macro
* Better document link argument of MANAGED_STACK_ADDRESS_BOEHM_GC_general_register_disappearing_link
* Better document missing debugging version for aligned allocation routines
* Better document preprocessor-related pointer checking routines in gc.h
* Better document why MANAGED_STACK_ADDRESS_BOEHM_GC_REVEAL_POINTER should be used with GC lock held
* Better separate libgc build- and use-time macro descriptions in README
* Build cord/de test by WCC_MAKEFILE on Win32
* Build extra/gc.c in digimars.mak
* Build with MANAGED_STACK_ADDRESS_BOEHM_GC_wcsdup support if wcslen exists
* Call MANAGED_STACK_ADDRESS_BOEHM_GC_init_parallel only from MANAGED_STACK_ADDRESS_BOEHM_GC_init (refactoring)
* Change MANAGED_STACK_ADDRESS_BOEHM_GC_Thread_Rep in pthread_support.h to match that of win32_threads.c
* Change NOT_MARKED value to non-zero (refactoring)
* Change default MANAGED_STACK_ADDRESS_BOEHM_GC_time_limit value from 50 to 15 ms

extra/gc.c
/*
 * Copyright (c) 1994 by Xerox Corporation.  All rights reserved.
 * Copyright (c) 1996 by Silicon Graphics.  All rights reserved.
 * Copyright (c) 1998 by Fergus Henderson.  All rights reserved.
 * Copyright (c) 2000-2009 by Hewlett-Packard Development Company.
 * All rights reserved.
 * Copyright (c) 2009-2018 Ivan Maidanski
 *
 * THIS MATERIAL IS PROVIDED AS IS, WITH ABSOLUTELY NO WARRANTY EXPRESSED
 * OR IMPLIED.  ANY USE IS AT YOUR OWN RISK.
 *
 * Permission is hereby granted to use or copy this program
 * for any purpose, provided the above notices are retained on all copies.
 * Permission to modify the code and to distribute modified code is granted,
 * provided the above notices are retained, and a notice that the code was
 * modified is included with the above copyright notice.
 */

/* This file could be used for the following purposes:          */
/* - get the complete GC as a single link object file (module); */
/* - enable more compiler optimizations.                        */

/* Tip: to get the highest level of compiler optimizations, the typical */
/* compiler options (GCC) to use are:                                   */
/* -O3 -fno-strict-aliasing -march=native -Wall -fprofile-generate/use  */

/* Warning: GCC for Linux (for C++ clients only): Use -fexceptions both */
/* for GC and the client otherwise MANAGED_STACK_ADDRESS_BOEHM_GC_thread_exit_proc() is not         */
/* guaranteed to be invoked (see the comments in pthread_start.c).      */

#ifndef __cplusplus
  /* static is desirable here for more efficient linkage.               */
  /* TODO: Enable this in case of the compilation as C++ code.          */
# define MANAGED_STACK_ADDRESS_BOEHM_GC_INNER STATIC
# define MANAGED_STACK_ADDRESS_BOEHM_GC_EXTERN MANAGED_STACK_ADDRESS_BOEHM_GC_INNER
                /* STATIC is defined in gcconfig.h. */
#endif

/* Small files go first... */
#include "../backgraph.c"
#include "../blacklst.c"
#include "../checksums.c"
#include "../gcj_mlc.c"
#include "../headers.c"
#include "../new_hblk.c"
#include "../obj_map.c"
#include "../ptr_chck.c"

#include "../allchblk.c"
#include "../alloc.c"
#include "../dbg_mlc.c"
#include "../finalize.c"
#include "../fnlz_mlc.c"
#include "../malloc.c"
#include "../mallocx.c"
#include "../mark.c"
#include "../mark_rts.c"
#include "../reclaim.c"
#include "../typd_mlc.c"

#include "../misc.c"
#include "../os_dep.c"
#include "../thread_local_alloc.c"

/* Most platform-specific files go here... */
#include "../darwin_stop_world.c"
#include "../dyn_load.c"
#include "../gc_dlopen.c"
#if !defined(PLATFORM_MACH_DEP)
# include "../mach_dep.c"
#endif
#if !defined(PLATFORM_STOP_WORLD)
# include "../pthread_stop_world.c"
#endif
#include "../pthread_support.c"
#include "../specific.c"
#include "../win32_threads.c"

#ifndef MANAGED_STACK_ADDRESS_BOEHM_GC_PTHREAD_START_STANDALONE
# include "../pthread_start.c"
#endif

/* Restore pthread calls redirection (if altered in             */
/* pthread_stop_world.c, pthread_support.c or win32_threads.c). */
/* This is only useful if directly included from application    */
/* (instead of linking gc).                                     */
#ifndef MANAGED_STACK_ADDRESS_BOEHM_GC_NO_THREAD_REDIRECTS
# define MANAGED_STACK_ADDRESS_BOEHM_GC_PTHREAD_REDIRECTS_ONLY
# include "gc/gc_pthread_redirects.h"
#endif

/* The files from "extra" folder are not included. */
